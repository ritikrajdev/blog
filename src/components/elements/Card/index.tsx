import React, { useState } from 'react';
import { PUT_BLOG } from '../../../constants/apiEndpoints/blog';
import { makeRequest } from '../../../utils/makeRequest';
import IconButton from '../IconButton';

import './Card.css';

export default function Card({
  id,
  image,
  date,
  reading_time: readingTime,
  title,
  claps,
  liked,
  description,
  imgAlt = '',
}: {
  id: string | number;
  image: string;
  date: string;
  reading_time: string;
  imgAlt?: string;
  title: string;
  claps: number;
  liked: boolean;
  description: string;
}) {
  const [isLiked, setIsLiked] = useState(liked);
  const [changeClap, setChangeClaps] = useState(false);

  const heartSrc = isLiked
    ? '/assets/icons/heart-red.svg'
    : '/assets/icons/heart-black.svg';

  const numClaps = claps + (changeClap ? 1 : 0);

  const _date = new Date(date);
  const formattedDate = _date.toLocaleDateString('en-GB', {
    day: 'numeric',
    month: 'short',
  });

  return (
    <div className='card'>
      <img src={image} alt={imgAlt} className='card-img' />

      <p className='card-meta-info card-padding'>
        <span>{readingTime}</span>
        <span>{formattedDate}</span>
      </p>

      <h3 className='card-title card-padding'>{title}</h3>
      <p className='card-content card-padding'>{description}</p>

      <div className='vertical-spacer'></div>

      <hr className='card-margin' />
      <div className='card-padding card-actions'>
        <div className='card-actions'>
          <IconButton
            iconPath='/assets/icons/clapping.svg'
            onClick={async () => {
              let newChangeClap = !changeClap;
              const newNumClaps = claps + (newChangeClap ? 1 : 0);

              try {
                await makeRequest(PUT_BLOG(id), {
                  data: {
                    claps: newNumClaps,
                  },
                });
              } catch (e) {
                // show some snacbar or anything else
                newChangeClap = changeClap;
              }

              setChangeClaps(newChangeClap);
            }}
          />
          <span>{numClaps}</span>
        </div>

        <div className='card-actions'>
          <IconButton
            iconPath={heartSrc}
            onClick={async () => {
              try {
                const response = await makeRequest(PUT_BLOG(id), {
                  data: {
                    liked: !isLiked,
                  },
                });
                setIsLiked(response.data.liked);
              } catch (e) {
                // show snackbar
              }
            }}
          />
        </div>
      </div>
    </div>
  );
}
