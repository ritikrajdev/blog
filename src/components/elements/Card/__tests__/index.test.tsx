import React from 'react';
import { describe, it } from '@jest/globals';

import Card from '..';
import { fireEvent, render, waitFor } from '@testing-library/react';
// import * as requests from '../../../../utils/makeRequest';

describe('Card', () => {
  const mockPost = {
    id: 1,
    date: '2nd Januray, 2018',
    // eslint-disable-next-line camelcase
    reading_time: '2 mins',
    title: 'The future of abstract art and the culture ...',
    description:
      'Create a blog post subtitle that summarizes your post in a few short, punchy sentences and entices your...',
    claps: 10,
    liked: false,
    image: 'abstract.png',
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correct data', () => {
    const screen = render(<Card {...mockPost} />);
    expect(screen.asFragment()).toMatchSnapshot();
  });

  it('should render with white heart if not liked', () => {
    const screen = render(<Card {...mockPost} />);
    expect(screen.getByAltText('/assets/icons/heart-black.svg')).toBeTruthy();
  });

  it('should change white heart to red on click', () => {
    jest.mock('../../../../utils/makeRequest', () => ({
      makeRequest: jest.fn().mockResolvedValue({
        data: {
          liked: true,
        },
      }),
    }));

    const screen = render(<Card {...mockPost} />);
    const whiteHeart = screen.getByAltText('/assets/icons/heart-black.svg');
    fireEvent.click(whiteHeart);
    waitFor(() => {
      const redHeart = screen.getByAltText('/assets/icons/heart-red.svg');
      expect(redHeart).toBeTruthy();
    });
  });

  it('should toggle heart on click', () => {
    jest.mock('../../../../utils/makeRequest', () => ({
      makeRequest: jest.fn().mockResolvedValue({
        data: {
          liked: true,
        },
      }),
    }));

    const screen = render(<Card {...mockPost} />);
    const whiteHeart = () =>
      screen.getByAltText('/assets/icons/heart-black.svg');
    fireEvent.click(whiteHeart());
    waitFor(() => {
      const redHeart = () => screen.getByAltText('/assets/icons/heart-red.svg');
      expect(redHeart).toBeTruthy();
      fireEvent.click(redHeart());
      waitFor(() => {
        expect(whiteHeart()).toBeTruthy();
      });
    });
  });

  it('should toggle claps on clapButton click', () => {
    jest.mock('../../../../utils/makeRequest', () => ({
      makeRequest: jest.fn().mockResolvedValue({}),
    }));

    const screen = render(<Card {...mockPost} />);
    const numClapsNode = screen.getByText(mockPost.claps);
    const clapsIconNode = screen.getByAltText('/assets/icons/clapping.svg');

    expect(numClapsNode.textContent).toBe(String(mockPost.claps));
    fireEvent.click(clapsIconNode);
    waitFor(() => {
      expect(numClapsNode.textContent).toBe(String(mockPost.claps + 1));
      fireEvent.click(clapsIconNode);
      waitFor(() => {
        expect(numClapsNode.textContent).toBe(String(mockPost.claps));
      });
    });
  });
});
