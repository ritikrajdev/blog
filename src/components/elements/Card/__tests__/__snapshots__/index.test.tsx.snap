// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render correct data 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <img
      alt=""
      class="card-img"
      src="abstract.png"
    />
    <p
      class="card-meta-info card-padding"
    >
      <span>
        2 mins
      </span>
      <span>
        2nd Januray, 2018
      </span>
    </p>
    <h3
      class="card-title card-padding"
    >
      The future of abstract art and the culture ...
    </h3>
    <p
      class="card-content card-padding"
    >
      Create a blog post subtitle that summarizes your post in a few short, punchy sentences and entices your...
    </p>
    <div
      class="vertical-spacer"
    />
    <hr
      class="card-margin"
    />
    <div
      class="card-padding card-actions"
    >
      <div
        class="card-actions"
      >
        <img
          alt="/assets/icons/clapping.svg"
          class="card-actions-icon"
          height="24"
          src="/assets/icons/clapping.svg"
          style="cursor: pointer;"
          width="24"
        />
        <span>
          10
        </span>
      </div>
      <div
        class="card-actions"
      >
        <img
          alt="/assets/icons/heart-black.svg"
          class="card-actions-icon"
          height="24"
          src="/assets/icons/heart-black.svg"
          style="cursor: pointer;"
          width="24"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
